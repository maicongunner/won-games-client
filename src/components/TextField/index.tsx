import { InputHTMLAttributes, useState } from 'react'
import * as S from './styles'

export type TextFieldProps = {
  onInput?: (value: string) => void
  label?: string
  initialValue?: string
  icon?: React.ReactNode
  iconPosition?: 'left' | 'right'
  disabled?: boolean
  messageError?: string
} & InputHTMLAttributes<HTMLInputElement>

const TextField = ({
  label,
  name,
  initialValue = '',
  onInput,
  icon,
  iconPosition = 'left',
  disabled = false,
  messageError = '',
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue)

  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = e.currentTarget.value
    setValue(newValue)

    !!onInput && onInput(newValue)
  }

  return (
    <S.Wrapper disabled={disabled} messageError={messageError}>
      {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      <S.InputWrapper>
        {!!icon && <S.Icon iconPosition={iconPosition}>{icon}</S.Icon>}
        <S.Input
          type="text"
          onChange={onChange}
          value={value}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </S.InputWrapper>
      {!!messageError && <S.messageError>{messageError}</S.messageError>}
    </S.Wrapper>
  )
}

export default TextField
