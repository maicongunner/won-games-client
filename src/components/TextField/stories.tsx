import { Story, Meta } from '@storybook/react/types-6-0'
import TextField, { TextFieldProps } from '.'

export default {
  title: 'TextField',
  component: TextField,
  args: {
    label: 'E-mail',
    labelFor: 'E-mail',
    id: 'Email',
    initialValue: '',
    placeholder: 'jhon.cage@gmail.com'
  },
  argTypes: {
    onInput: { action: 'changed' }
  }
} as Meta

export const Default: Story<TextFieldProps> = (args) => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
)
