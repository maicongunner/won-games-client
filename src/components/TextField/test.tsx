import { Email } from '@styled-icons/material-outlined'
import { screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderWithTheme } from 'utils/tests/helpers'

import TextField from '.'

describe('<TextField />', () => {
  it('should render correctly', () => {
    const { container } = renderWithTheme(<TextField label="text" />)

    expect(container).toMatchSnapshot()
  })

  it('should render with label', () => {
    renderWithTheme(<TextField label="text" />)

    expect(screen.getByText(/text/i)).toBeInTheDocument()
  })

  it('should render without label', () => {
    renderWithTheme(<TextField />)

    expect(screen.queryByText(/text/i)).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder="text placeholder" />)

    expect(screen.getByPlaceholderText(/text/i)).toBeInTheDocument()
  })

  it('changes its value when typing', async () => {
    const onInput = jest.fn()
    renderWithTheme(
      <TextField
        label="textField"
        labelFor="textField"
        id="textField"
        onInput={onInput}
      />
    )

    const input = screen.getByRole('textbox')
    const text = 'my new text'

    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInput).toHaveBeenCalledTimes(text.length)
    })
    expect(onInput).toHaveBeenCalledWith(text)
  })

  it('should render with icon', () => {
    renderWithTheme(
      <TextField icon={<Email data-testid="icon" />} label="text" />
    )

    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })

  it('should render with icon on the left side', () => {
    renderWithTheme(
      <TextField icon={<Email data-testid="icon" />} label="text" />
    )

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      order: 0
    })
  })

  it('should render with icon on the right side', () => {
    renderWithTheme(
      <TextField
        icon={<Email data-testid="icon" />}
        label="text"
        iconPosition="right"
      />
    )

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({
      order: 1
    })
  })
})
