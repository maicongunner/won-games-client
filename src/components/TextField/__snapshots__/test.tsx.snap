// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render correctly 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c1:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c2 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c0 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

<div>
  <div
    class=""
  >
    <label
      class="c0"
      for=""
    >
      text
    </label>
    <div
      class="c1"
    >
      <input
        class="c2"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`
